src/libraries/System.Text.Json/tests/Common/PropertyVisibilityTests.cs

```diff
@@ -3028,5 +3225,84 @@ public class Implementation : IJoinInterface
+
+            collection = await Serializer.DeserializeWrapper<CollectionWithPrivateElementType>(json);
+            collection.Validate();
+        }
+
+        public class CollectionWithPrivateElementType : ICollection<PrivateEnum>
+        {
+            private readonly ICollection<PrivateEnum> _values = new List<PrivateEnum>();
+
+            public static CollectionWithPrivateElementType CreatePopulatedInstance()
+                => new CollectionWithPrivateElementType { _values = { PrivateEnum.A, PrivateEnum.B, PrivateEnum.C } };
+
+            public void Validate() => Assert.Equal(new[] { PrivateEnum.A, PrivateEnum.B, PrivateEnum.C }, this);
+            public string GetExpectedJson() => "[0,1,2]";
+
+            int ICollection<PrivateEnum>.Count => _values.Count;
+            bool ICollection<PrivateEnum>.IsReadOnly => _values.IsReadOnly;
+            void ICollection<PrivateEnum>.Add(PrivateEnum item) => _values.Add(item);
+            void ICollection<PrivateEnum>.Clear() => _values.Clear();
+            bool ICollection<PrivateEnum>.Contains(PrivateEnum item) => _values.Contains(item);
+            void ICollection<PrivateEnum>.CopyTo(PrivateEnum[] array, int arrayIndex) => _values.CopyTo(array, arrayIndex);
+            IEnumerator<PrivateEnum> IEnumerable<PrivateEnum>.GetEnumerator() => _values.GetEnumerator();
+            IEnumerator IEnumerable.GetEnumerator() => _values.GetEnumerator();
+            bool ICollection<PrivateEnum>.Remove(PrivateEnum item) => _values.Remove(item);
+        }
+
+        [Fact]
+        public async virtual Task TestDictionaryWithPrivateKeyAndValueType()
+        {
+            // The reflection-based serializer supports dictionaries whose key/value types are private
+
+            DictionaryWithPrivateKeyAndValueType collection = DictionaryWithPrivateKeyAndValueType.CreatePopulatedInstance();
+
+            string json = await Serializer.SerializeWrapper(collection);
+            Assert.Equal(collection.GetExpectedJson(), json);
+
+            collection = await Serializer.DeserializeWrapper<DictionaryWithPrivateKeyAndValueType>(json);
+            collection.Validate();
+        }
+
+        public class DictionaryWithPrivateKeyAndValueType : IDictionary<PrivateEnum, PrivateEnum>
+        {
+            private readonly IDictionary<PrivateEnum, PrivateEnum> _values = new Dictionary<PrivateEnum, PrivateEnum>();
+
+            public static DictionaryWithPrivateKeyAndValueType CreatePopulatedInstance()
+                => new DictionaryWithPrivateKeyAndValueType { _values = { [PrivateEnum.A] = PrivateEnum.B, [PrivateEnum.B] = PrivateEnum.C } };
+
+            public void Validate() => Assert.Equal(new KeyValuePair<PrivateEnum, PrivateEnum>[] { new(PrivateEnum.A, PrivateEnum.B), new(PrivateEnum.B, PrivateEnum.C) }, this);
+            public string GetExpectedJson() => """{"A":1,"B":2}"""; // cf. https://github.com/dotnet/runtime/issues/87129
```

Comment:
IIRC this is by-design - https://github.com/dotnet/runtime/issues/87129#issuecomment-1577298024.